### ISSUE TITLE
[osx_defaults] Add dict/dict-add support

### ISSUE THREAD

**Issue by claco** [ansible/ansible#24028]
_From @claco on February 21, 2016 16:31_

##### Issue Type:
- Feature Idea
##### Plugin Name:

osx_defaults
##### Ansible Version:

```
ansible 2.0.0.2
  config file = /Users/claco/Projects/claco.network/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

```
[defaults]
ask_sudo_pass = True
fact_caching = jsonfile
fact_caching_connection = ./facts
fact_caching_timeout = 86400
hash_behaviour: merge
inventory = ./inventory
roles_path = ./roles
vault_password_file = ~/.ansible_vault_password
```
##### Environment:

OSX El Cap 10.11
##### Summary:

The osx_defaults module doesn't yet support dict and dict-add methods for changing/adding defaults that are contained within values that are dicts.
##### Steps To Reproduce:

As an example, I'm configuring defaults for SizeUp on a new install. Many of the settings are container in dicts for each group of key commands. This command, for example, configures the combo code for the 'move window to lower left' command:

```
defaults write com.irradiatedsoftware.SizeUp. "Lower Left" -dict-add ComboCode 125
```

Because there is no dict/dict-add support, I have to fall back to using bash/shell commands rather than using the module.
##### Expected Results:

As an example, I would have expected to see something like this:

```
- name: Add Default Lower Left
  osx_defaults: domain=com.irradiatedsoftware.SizeUp key="Lower Left", value: {ComboCode: 125} state=present
```
##### Actual Results:

What actually happened?

```
failed: [mqr13tg8wl] => (item={u'type': u'dict', u'key': u'Lower Left', u'value': u'ComboFlags 126'}) => {"failed": tr
ue, "item": {"key": "Lower Left", "type": "dict", "value": "ComboFlags 126"}, "msg": "value of type must be one of: ar
ray,bool,boolean,date,float,int,integer,string, got: dict"}
```

_Copied from original issue: ansible/ansible-modules-extras#1705_

**Comment by claco** [comment]
For whoever reviews this, I'm working on a PR to implement basic dict/dict-add functionality. Should have a first pass this week.

**Comment by claco** [comment]
Ok, another round of refactoring complete so I removed the WIP. Because `defaults read` on a dictionary key type yields quoted values that don't match the underlying type ("0.5" returned, but type is "int" in plist editor), I've changed the value for `dict` types to be more verbose:

You can pass either a single dict, which might be convenient for `dict-add` situations, or an array of `dict` values if you're describing an entire `dict` (replacing) entry:

```
- osx_defaults:
    domain: com.irradiatedsoftware.SizeUp
    key: Lower Left
    type: dict
    value: { key: ComboCode, type: int, value: 125 }
    dict_add: True

- osx_defaults:
    domain: com.irradiatedsoftware.SizeUp
    key: Lower Left
    type: dict
    value:
      - { key: ComboCode, type: int, value: 125 }
      - { key: ComboFlags, type: int, value: 1835008 }
```

**Comment by claco** [comment]
_From @danbohea on February 26, 2016 18:44_

+1 for this feature, just bumped into a need for it today.

**Comment by claco** [comment]
_From @ansibot on August 4, 2016 20:28_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by claco** [comment]
_From @ansibot on September 13, 2016 3:34_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by claco** [comment]
_From @ansibot on October 6, 2016 18:8_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by claco** [comment]
_From @ansibot on December 7, 2016 15:52_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
@claco Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by daften** [comment]
Hey @claco  Thanks for this, can you update the issue? :)

**Comment by ansibotdev** [comment]
@claco You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dixhuit** [comment]
This thread makes me want to cry. There is plenty of info here yet the official responses are just nit-picks automated guff.

**Comment by daften** [comment]
True, but to be fair, if this stuff doesn't happen, a lot of requests will just remain open forever, and there isn't even a possibility for decent overview.
